Step5 - 로또(수동)

mins99 opened the pull request:
안녕하세요 리뷰어님!
step4의 피드백을 반영하고 step5 구현하여 로또 미션 마지막 리뷰 요청 드립니다 🥳
step5가 아무래도 마지막이기도 해서 구현 요구사항 뿐만 아니라 구조에 있어 고민을 더 많이 했던 것 같습니다!

구현에 있어 고민했던 부분은 코멘트로 남겨두겠습니다 ㅎㅎ
리뷰어님의 좋은 피드백 덕에 보람찬 시간 보낼 수 있었습니다. 정말 감사했습니다!

\+ 추가로 [지속가능한 SW 개발을 위한 코드리뷰](https://www.youtube.com/watch?v=ssDMIcPBqUE) 영상을 본적이 있었는데요 
1) 리뷰를 받는 사람이 구현에 있어 고민한 부분 
2) 이부분을 더 중점적으로 봐주었으면 좋겠다 하는 부분  

에 대해 코멘트를 남기는 것도 리뷰어에게 좋다고 하셔서 저도 코멘트를 남기기 시작한건데.. 리뷰를 하시는 입장에서 어떠셨는지 여쭤봐도 될까요?!
다음 미션 진행에 참고 하고 싶어서 의견 여쭤봅니다 ㅎㅎ (너무 개인적인 질문인거면 DM으로 답변 주셔도 됩니다..!)

아무쪼록 마지막 step5 마무리까지 잘 부탁드립니당 🙇🏻

mins99 committed:
docs(README): 5단계 - 로또(수동) 기능 요구 사항 정리

mins99 committed:
refactor(GameResult): step4 calculateRank 피드백 반영

- 로또 번호 일치 갯수(countOfMatch)를 winningLotto에서 계산하도록 로직 수정

mins99 committed:
refactor(WinningLotto): 생성자 추가

- 테스트에서 입력을 편하게 받기 위해 추가함

mins99 committed:
refactor(GameResult): mapToInt -> mapToLong 변경

- 4000000000원 같이 큰 수인 경우 연산 값이 안맞아서 수정

mins99 committed:
feat(LottoCount): 로또 구매 장수에 대한 객체 추가

- 기존 Price 객체의 역할은 Coin이 가지도록 함

mins99 committed:
feat(Price): 객체 삭제

- 기존 Price 객체의 역할은 Coin이 가지도록 함

mins99 committed:
feat: 로또(수동) 구현

- 구매 장수에 대한 로직은 수동과 자동으로 각각 분리
- 구매 금액에 대한 정보는 Coin이, 구매 장수에 대한 정보는 LottoCount 가 가지고 있음

mins99 committed:
test(CoinTest): 구매 금액에 대한 테스트 추가

mins99 committed:
test(LottoCountTest): 구매 장수에 대한 테스트 추가

mins99 committed:
test(GameResult): 게임 결과 테스트 로직 개선

mins99 committed:
test(LottoGameTest): 로또 게임 메인 로직 테스트 개선 및 수동 로또 관련 테스트 추가

mins99 committed:
docs(README): 5단계 - 로또(수동) 구현 사항 체크

mins99 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -23,6 +24,8 @@ public void startGame() {
     private void play(LottoGame game) {
         insertMoney(game);
 
+        purchaseLotto(game);
+
         inputWinnerNumbers(game);
         inputBonusNumbers(game);
 
```
`LottoGame` 내의 객체에서 유효성 검증을 진행하다 보니 모든 메서드에 `LottoGame`를 파라미터로 전달하는 구조가 되었습니다;; 
util로 input 데이터에 대한 검증 로직을 새로 생성해서 처리하는 방식이 나을까요? 
(PR 요청 드리고 떠오른 아이디어라 뒤늦게 코멘트로 답니다 ㅎㅎ;)

oper912 replied:
음 개인적으로는 입력받은 문자를 숫자형으로 변환해서 넘겨주는 것 정도는 InputView에서 해도 괜찮다고 생각하는데요
실제로 프론트에서 API스펙에 맞게 원하는 값을 전달받았다고 생각해주시면 좋을 것 같아요

질문주신 `LottoGame` 은 현재는 멀티모듈로 구조를 잘 못 잡았을때 common 모듈만 덩치가 점점 커지는 느낌이긴하네요
로또를 진행하는데 있어서 모든 책임이 `LottoGame` 에게 할당되어 버리는 건 아닌가 고민해 볼 필요가 있겠네요

oper912 replied:
`LottoGame` 이 가지는 책임이나 알고 있는 정보를 조금 줄여보면 좋을 것 같습니다.
간단하게 생각해봤을 때는 아래와 같은 흐름으로만 진행되어도 객체간의 역할이 어느정도 구분되었을 거 같긴하네요

1. 구입금액을 입력받아서 `Coin` 객체를 생성한다. (잘못된 구입금액의 validation은 Coin의 책임)
2. 수동으로 구매할 로또 수, 번호를 입력받아서 `LottoGame`에 전달한다 (수동 및 자동 로또 번호 생성)
3. 지난주 당첨번호, 보너스볼을 입력받아서 `WinningLotto`생성
4. `GameResult` 에 `LottoGame` (구입 금액과 구입된 로또 정보) 와 `WinningLotto` (당첨번호) 를 전달해서 결과를 정보를 생성한다.
5. `ResultView` 에 `GameResult` 를 전달해서 결과를 출력한다.

mins99 replied:
피드백 주신대로 수정을 해보니까 LottoGame 로직이 간단해졌네요. 
테스트에 있어서도 복잡했었는데 많이 간단해졌습니다!
감사합니다 : )

mins99 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -34,9 +37,27 @@ private void insertMoney(LottoGame game) {
+    private void purchaseLotto(LottoGame game) {
+        int countOfManualLotto = inputManualLottoCount(game);
+
+        while (countOfManualLotto != game.getUserLotto().size()) {
+            game.inputManualLottoNumber(inputView.inputManualLottoNumber(game.getUserLotto().size() == ZERO));
+        }
```
`game.getUserLotto().size() == ZERO` 같은 경우 
while문 바로 위에 `inputView.inputManualLottoNumber()` 를 호출하여 구현할 수도 있지만 
수동 로또가 0개인 경우에도 `수동으로 구매할 번호를 입력해 주세요` 문구가 출력되어서 이를 처리해보고자 넣게 되었습니다

oper912 replied:
`game.getUserLotto().size() == ZERO ` 와 같은 경우 객체에서 정보를 꺼내와서 무언가를 하기보다는 
`현재 userLotto의 size가 0개와 같는지` 질문하듯이 메시지를 전달하는 방식으로 구현해보면 어떨까요??

mins99 replied:
`객체에게 메시지를 던진다` 계속 연습하려고는 하는데 어렵네요 ㅠㅠ
`game.isSameSizeOfUserLotto(countOfManualLotto)` 와 같은 형식으로 수정해보겠습니당 😀

mins99 commented on src/main/java/lotto/model/Coin.java:
```diff
@@ -2,25 +2,32 @@
         }
     }
+
+    private void minimumPriceCheck() {
+        if (deposit < LOTTO_PRICE) {
+            throw new IllegalArgumentException("최소 구입금액은 " + LOTTO_PRICE + "원 입니다.");
+        }
+    }
```
`Coin`이 최소 금액(LOTTO_PRICE)을 가지고 있도록 변경 하여 최소 구입금액의 유효성 검증 역시 `Coin`이 하도록 변경했습니다!
`Coin`의 경우 금액과 관련된 로직을 처리하며,
`LottoCount`(구 Price)의 경우 수량과 관련된 로직을 처리하도록 구현했습니다

oper912 replied:
이전보다 훨씬 객체지향에 대한 개념이 잘 잡히신 것 같아요 👍

mins99 commented on src/main/java/lotto/model/GameResult.java:
```diff
@@ -10,10 +10,7 @@ public class GameResult {
-                .filter(winningLotto::contains)
-                .count();
-
+        int countOfMatch = winningLotto.countOfMatchNumber(userLotto);
```
https://github.com/next-step/java-lotto-pro/pull/534#discussion_r875371837 와 관련하여
winningLotto에서 모든 비교로직을 하도록 수정해보았습니다
(훨씬 간결해진것 같아서 좋습니당 🥰)

mins99 commented on src/main/java/lotto/model/GameResult.java:
```diff
@@ -30,6 +27,6 @@ public Map<Rank, Integer> gameResult() {
 
     public double benefitResult(double deposit) {
-        return gameResult.stream().mapToInt(Rank::getWinningMoney).sum() / deposit;
+        return gameResult.stream().mapToLong(Rank::getWinningMoney).sum() / deposit;
```
1등이 많이 나오는(?!) 경우에는 int의 범위를 넘어가서 확장성(?)을 고려한 mapToLong 으로 변경하였습니다. 💸

mins99 commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -5,6 +5,10 @@ public class WinningLotto {
     private final Lotto lotto;
     private final Number bonusNo;
 
+    public WinningLotto(String inputs, String bonusNo) {
+        this(new Lotto(new InputNumberGenerator(inputs)), new Number(bonusNo));
+    }
```
로또 피드백 영상에서 포비께서 원시값과 문자열을 포장 함으로서 테스트 코드가 길어지는 경우의 대처법(?)을 알려주셔서 사용해보았습니다

oper912 replied:
음.. 혹시 생성자가 여러개일 경우에 메인 생성자를 하나두고 객체 내부에서 서브 생성자들이 메인 생성자들을 호출하도록 하라는 말씀이 아니셨을까요..?

mins99 replied:
여러개의 서브 생성자(public)를 통해 최종적으로 메인 생성자(private)를 호출하는 구조를 말씀하시는 걸까요?
영상에서는 생성자가 모두 public이라 파라미터가 다양한 public 생성자를 구현하려는 의도로 해석했었습니다
우선은 LottoGame의 역할을 줄이면서 WinningLotto를 String으로만 생성하는 정적 팩토리 메서드로 수정해볼까 합니다 ㅎㅎ;

mins99 commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -18,6 +22,12 @@ public boolean contains(Number number) {
         return lotto.contains(number);
     }
 
+    public int countOfMatchNumber(Lotto userLotto) {
+        return (int) userLotto.getNumbers().stream()
+                .filter(this.lotto::contains)
+                .count();
+    }
```
https://github.com/next-step/java-lotto-pro/pull/534#discussion_r875371837
말씀 주셨던 부분이 다음과 같이 WinningLotto에서 처리하는 로직이 맞을까요~?

oper912 replied:
네 맞습니다~

mins99 reviewed (commented):
생각보다 고민한 부분이 있어 그런지 코드 수정내역이 많네요 ㅎㅎ;
코멘트 달면서 저도 또 다시 코드를 보게 되어 개인적으로는 좋은것 같습니다!

oper912 reviewed (changes requested):
안녕하세요 민성님 :)

질문 주신 부분 우선 답변 드리면 미션을 진행하시면서 구현에 있어서 고민한 부분이나 또는 중점적으로 같이 의견을 나누고 싶은 부분이 있으면 말씀해주시면 당연히 너무 좋습니다 
미션 진행해주신 부분에 대해서 피드백 드릴 때 참고해서 드릴 수 있을 것 같아요 😀

하지만 리뷰어분들이 드리는 피드백이 무조건 정답은 아니라는 점만 알아주셨으면 좋겠어요
같이 의견을 나누면서 조금 더 나은 방향을 찾아간다고 생각해주시면 좋을 것 같구 정답을 바라는 질문보다는 고민되는 부분에 대해서 함께 고민할 수 있는 방향이면 매우 좋을 것 같아요 👍

진행해주신 미션에 대해서 아쉬운 부분들이 있어서 코멘트 남겨놓았는데요 
`LottoGame` 객체가 가진 너무 과도한 책임에 대해서 조금 더 고민해보고 리뷰 재요청 부탁드립니다 🙏

mins99 committed:
refactor(LottoGame, GameResult): LottoGame이 가진 지나친 책임을 분리

- WinningLotto 는 Controller 에서 생성된 객체를 GameResult 로 전달하도록 수정
- GameResult 는 LottoGame 과 WinningLotto 를 전달 받아 결과 및 수익률을 생성하도록 수정
- 유효하지 않은 값이 입력되는 경우의 예외처리를 controller 에서 출력하고 끝내도록 수정

mins99 committed:
fix(GameResult): 카운트 타입을 int -> long으로 변경

mins99 committed:
fix(GameResult): GameResult 가 가지는 객체 변경

- 순위 리스트인 rankList 와 유저의 로또구입금액인 userDeposit 을 가지도록 수정

mins99 committed:
fix(Coin): deposit 계산 로직 변경

- 유저가 1000원 이상의 금액을 입력시 천원 이하는 미사용

mins99 committed:
test(Coin): deposit 계산 로직 테스트 추가

mins99 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -1,60 +1,48 @@
     public void startGame() {
-        LottoGame game = new LottoGame();
-        play(game);
+        try {
+            LottoGame game = new LottoGame();
+            play(game);
+        } catch (Exception e) {
+            resultView.printErrorMessage(e.getMessage());
+        }
```
기존 LottoGame이 예외 처리를 하는 역할도 담당하고 있었는데, 과도한 책임인것 같아서 우선 오류에 대한 예외 처리도 컨트롤러에서 하도록 수정해보았습니다. 
도메인에서는 예외를 던지기만 하고 컨트롤러에서는 예외 처리를 하는 식으로 분리 했습니다.

mins99 commented on src/main/java/lotto/model/Coin.java:
```diff
@@ -2,18 +2,20 @@
 
 public class Coin {
     private final int deposit;
+    private static final int LOTTO_PRICE = 1000;
 
     public Coin(String money) {
         invalidMoneyCheck(money);
-        deposit = Integer.parseInt(money);
+        this.deposit = Integer.parseInt(money) / LOTTO_PRICE * LOTTO_PRICE;
+        minimumPriceCheck();
     }
```
백 원 단위 이하의 금액은 구입 금액에 계산하지 않도록 수정했습니다.

mins99 commented on src/main/java/lotto/model/GameResult.java:
```diff
@@ -7,29 +7,41 @@
 
 public class GameResult {
-    private final List<Rank> gameResult = new ArrayList<>();
+    private final List<Rank> rankList;
+    private final double userDeposit;
+    private static final double BENEFIT_REFERENCE_VALUE = 1;
```
LottoGame이 가지고 있던 책임을 분리하고 나니 private 메서드로 처리가 가능한 메서드가 많아지기도 했고, 실제 테스트시에 계속 gameResult 객체에 전달해야 하는 것들이 많았었는데 그런 부분도 어느 정도 처리가 가능해지는 느낌을 많이 받을 수 있었습니다! 😍

oper912 replied:
개인적으로 로또 미션은 
TDD를 통해 테스트코드 작성한 구조로 설계를 하면서 
협력에 필요한 책임에 따라서 적절하게 역할을 나누고 객체들에게 부여하는 객체지향적인 사고능력이 핵심이라고 생각하는데 
5단계 미션을 진행하시면서 얻어가는 부분이 있으신거 같아서 다행입니다 👍

mins99 reviewed (commented):
코멘트를 다 추가하고 리뷰 요청을 드렸어야 했는데 먼저 리뷰 요청을 해버렸네요 😅
`LottoGame` 의 역할을 분리하면서 기존에 고민했던 Controller 로직도 같이 수정을 하게 되었습니다.
처음에는 Controller가 `LottoGame` 만 가지고 있게 해보자는 생각으로 구현을 진행했었는데,
그 부분에 있어 기능이 추가되는 측면에서 코드를 더 복잡하게 만드는 요소로 작용한것 같습니다..😭
각 도메인들이 역할을 너무 많이 가지지 않도록 신경쓰면서 수정해보았습니다. 확인 부탁드립니다~!

oper912 reviewed (approved):
안녕하세요 민성님 :) 
피드백 잘 반영해주셨네요! 이전보다 훨씬 객체지향적으로 구조가 잘 잡힌 것 같아요 💯

미션 진행하시는 동안 민성님이 고민되는 부분, 로직을 구현한 근거등을 말씀해주셔서 너무 좋았고
또 피드백 드리지 않았던 부분에 대해서도 계속해서 더 나은 방향으로 고민하는 부분이 매우 인상 깊었습니다
부족한 리뷰였지만 조금이라도 얻어가시는게 있었으면 좋겠네요 🙏

로또 미션은 여기서 마무리하도록 하겠습니다 
남은 미션들도 지금 처럼만 해주신다면 충분히 잘 해주실 것 같아요 😀
고생 많으셨어요! 화이팅입니다! 💪
